### PR TITLE
Job::enqueue_sync_to_index: Only insert new jobs if no similar unlocked jobs exist

### PR DESCRIPTION
**Pull request by Turbo87**
This PR implements the equivalent of:

```sql
INSERT INTO background_jobs (job_type, data, priority) (
    SELECT 'sync_to_git_index', '{"krate":"serde"}'::jsonb, -50
    WHERE NOT EXISTS (
        SELECT id FROM background_jobs WHERE job_type = 'sync_to_git_index' AND data = '{"krate":"serde"}'::jsonb FOR UPDATE SKIP LOCKED
    )
) UNION ALL (
    SELECT 'sync_to_sparse_index', '{"krate":"serde"}'::jsonb, -50
    WHERE NOT EXISTS (
        SELECT id FROM background_jobs WHERE job_type = 'sync_to_sparse_index' AND data = '{"krate":"serde"}'::jsonb FOR UPDATE SKIP LOCKED
    )
);
```

This should ensure we don't enqueue a bunch of duplicate jobs if someone decides to yank/unyank a lot of versions of the same crate in a short amount of time.

### CODE REVIEW

File: src/background_jobs.rs [hunk 1]
```diff
@@ -108,24 +110,45 @@ impl Job {
     ) -> Result<(), EnqueueError> {
         use crate::schema::background_jobs::dsl::*;
 
+        let find_similar_jobs_query = |job: &Job| {
+            let query = background_jobs
+                .select(id)
+                .filter(job_type.eq(job.as_type_str()))
+                .filter(data.eq(job.to_value()?))
+                .filter(priority.eq(PRIORITY_SYNC_TO_INDEX))
```
Comment (LawnGnome) [inline src/background_jobs.rs:118]: The priority clause isn't in the hand written queries in the PR description — I assume that's an intentional change when you were converting this to the Diesel DSL syntax?

**Review by LawnGnome — COMMENTED** [review]
As we talked about last week, this definitely needs some inline commentary before we merge it, but I think this is good otherwise pending tests.

File: src/background_jobs.rs (hunk 1 above)
Comment (Turbo87) [reply src/background_jobs.rs:120]: yeah, I figured if the `priority` was different there might be a reason for duplicating the job 🤷‍♂️

**Comment by Turbo87** [comment]
> needs some inline commentary before we merge it

✅  let me know if this is sufficient or if you need more :)

> pending tests

I've tried to write tests for this, but due to https://github.com/rust-lang/crates.io/blob/9189d1a76196d7ec772baffee5c8500ff45593c5/src/tests/util/test_app.rs#L97-L99 this has proven quite difficult. Without a major refactoring of our test suite I'm not sure how to write proper tests for this :-/

**Review by LawnGnome — APPROVED** [review]
> white_check_mark let me know if this is sufficient or if you need more :)

Nope, that looks good to me. Thanks!

Given the test difficulties, I'm happy going ahead with this as is for now — maybe we just create a follow up issue noting that we'd like to have test coverage here, and maybe some bright person will jump in, or one of us will have a sudden brainwave.

**Comment by Turbo87** [comment]
> maybe we just create a follow up issue noting that we'd like to have test coverage here, and maybe some bright person will jump in, or one of us will have a sudden brainwave.

I've added a doc comment to the function, including a warning about the function not having any explicit tests.

**Comment by Turbo87** [comment]
I've just tested this out on staging and subsequently fixed a small calculation bug regarding the `skipped_jobs_count` value, but generally this appears to work as intended now.

**Pull request merged**